### PR TITLE
Enabled Client Side Caching #167

### PR DESCRIPTION
**Pull request by archiejain28**
Hi @anmode,

As discussed, I have enable caching at client end and also removed server side caching (getStaticProps).

Kindly review the PR. 

#167

### CODE REVIEW

**Comment by archiejain28** [comment]
Hi @anmode ,
I have added Spinner while loading the data. I added it as Spinner component. So it can be used in any other place also.

File: components/Spinner.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+const Spinner =()=>{
+    return (
+    <div class="h-screen bg-white">
+    <div class="flex justify-center items-center h-full">
+      <img class="h-16 w-16" src="https://icons8.com/preloaders/preloaders/1488/Iphone-spinner-2.gif" alt="" />
+    </div>
+    </div>)
+}
+
+export default Spinner;
```
Comment (anmode) [inline components/Spinner.js:13]: Thanks for this component! Thinking to keep it in a `basic` folder named in the component folder!

File: pages/mentors.page.js [hunk 2]
```diff
@@ -1,21 +1,31 @@
...
 const Header = dynamic(() => import("../components/Header"));
 const SimpleBanner = dynamic(() => import("../components/SimpleBanner"));
 
-function Mentors({ mentorsData }) {
+function Mentors() {
+
+  const url = `${process.env.NEXT_PUBLIC_BACKEND_URL}/api/mentors/mentorLists`;
+  const data = useApi(url);
+  const mentorsData= data.filter((mentor) =>mentor.verified === true && mentor.token === "mentorIsVerified");
+
```
Comment (anmode) [inline pages/mentors.page.js:15]: for now could you please comment it out!

File: pages/mentors.page.js [hunk 3]
```diff
@@ -33,18 +43,3 @@ function Mentors({ mentorsData }) {
...
-  return {
-    props: {
-      mentorsData: data.filter(
-        (mentor) =>
-          mentor.verified === true && mentor.token === "mentorIsVerified"
-      ),
-    },
-    revalidate: 20, // revalidate the data every 20 seconds
-  };
-};
```
Comment (anmode) [inline pages/mentors.page.js:50]: revert this, please!

**Review by anmode — APPROVED** [review]
Thanks @archiejain28 for two components
1. Spinner
2. client-side caching.
I think we should not use as of now.... I have see why first call api took lots of time to query data from DB
1. Mongodb server ( shared)

**Review by anmode — COMMENTED** [review]
```
import React, { useState, useEffect, useRef, useCallback } from "react";
import axios from "axios";

export const useApi = (url) => {
  const [apidata, setApiData] = useState([]);
  const isCachePresent = useRef(sessionStorage.getItem('isCachePresent') === 'true');

  const fetchData = useCallback(async (url) => {
    try {
      const { data } = await axios.get(url);
      if (data) {
        setApiData(data);
      }
    } catch (err) {
      console.error('Error fetching data:', err);
      // Handle the error (e.g., display an error message)
    }
  }, []);

  const startCaching = useCallback(async () => {
    const newCache = await caches.open('new-cache');
    await newCache.add(url);
    if (newCache) {
      const res = await caches.has('new-cache');
      if (res) {
        sessionStorage.setItem('isCachePresent', 'true');
        isCachePresent.current = true;
      }
    }
  }, [url]);

  const datafromCache = useCallback(async () => {
    try {
      let cacheResponse = await caches.match(url);
      if (cacheResponse) {
        let body = await cacheResponse.json();
        setApiData(body);
        isCachePresent.current = true;
      } else {
        fetchData(url);
      }
    } catch (err) {
      console.error('Error retrieving data from cache:', err);
      // Handle the error (e.g., display an error message)
    }
  }, [url, fetchData]);

  useEffect(() => {
    if (apidata.length === 0 && !isCachePresent.current) {
      fetchData(url);
      startCaching();
    } else {
      datafromCache();
    }
  }, [url, apidata.length, isCachePresent.current, fetchData, startCaching, datafromCache]);

  return apidata;
};
```

**Comment by anmode** [comment]
@archiejain28 ??

**Review by anmode — APPROVED** [review]
cool!! thanks @archiejain28

**Comment by anmode** [comment]
Some error:
1. Run `npm run build` - you will see error
```


./pages/mentors.page.js
--
22:04:22.210 | Module not found: Can't resolve './customHook/useApi'


``
2. Fix the formatting- `npx prettier --write .`

**Pull request merged**